templates/shift-scheduling/main.go

```diff
@@ -1,370 +1,248 @@
-			changes = append(changes,
-				shifts.Remove(morning, []int{int(v.Worker)}),
-				shifts.Remove(day, []int{int(v.Worker)}),
-				shifts.Remove(night, []int{int(v.Worker)}),
-			)
+	demandCovering := map[int][]Assignment{}
+	for _, demand := range input.RequiredWorkers {
+		demandCovering[demand.RequiredWorkerID] = []Assignment{}
+		for i, potentialAssignment := range potentialAssignments {
+			if (potentialAssignment.Start.Before(demand.Start.Time) || potentialAssignment.Start.Equal(demand.Start.Time)) &&
+				(potentialAssignment.End.After(demand.End.Time) || potentialAssignment.End.Equal(demand.End.Time)) {
+				potentialAssignments[i].DemandsCovered = append(potentialAssignments[i].DemandsCovered, demand)
+				demandCovering[demand.RequiredWorkerID] = append(demandCovering[demand.RequiredWorkerID], potentialAssignment)
+			}
 		}
 	}
-	schedule = schedule.Apply(changes...)
 
-	// Now we define our objective value. This is the quantity we try to
-	// minimize. It is the weighted sum of workers and happiness. Happiness is
-	// negative as we want to maximize it while minimizing the worker count.
-	workerCountImportance := 10
-	schedule = schedule.Value(func(s store.Store) int {
-		return workerCountImportance*workerCount.Get(s) - happiness.Get(s)
-	})
-
-	// Now we define our constraint containing a propagator. This will remove
-	// workers from domains based upon the workers in other domains.
-	constraint := newBreakConstraint(shifts, 2)
-
-	// We invoke the propagator to propagate the initial values.
-	schedule.Propagate(constraint.propagate)
-
-	if constraint.isProvenInfeasible(schedule) {
-		return nil, fmt.Errorf("input can not be solved")
+	m := mip.NewModel()
+	m.Objective().SetMinimize()
+
+	x := model.NewMultiMap(
+		func(...Assignment) mip.Bool {
+			return m.NewBool()
+		}, potentialAssignments)
+
+	underSupplySlack := model.NewMultiMap(
+		func(demand ...RequiredWorker) mip.Float {
+			return m.NewFloat(0, float64(demand[0].Count))
+		}, input.RequiredWorkers)
+
+	overSupplySlack := model.NewMultiMap(
+		func(demand ...RequiredWorker) mip.Float {
+			return m.NewFloat(0, math.MaxFloat64)
+		}, input.RequiredWorkers)
+
+	for _, demand := range input.RequiredWorkers {
+		demandCover := demandCovering[demand.RequiredWorkerID]
+		// We need to cover all demands
+		coverConstraint := m.NewConstraint(mip.Equal, float64(demand.Count))
+		coverConstraint.NewTerm(1.0, underSupplySlack.Get(demand))
+		coverConstraint.NewTerm(-1.0, overSupplySlack.Get(demand))
+		coverPerDriver := map[int]mip.Constraint{}
+		for _, assignment := range demandCover {
+			if constraint, ok := coverPerDriver[assignment.Worker.ID]; !ok {
+				// When two potential assignments of a driver overlap, only one
+				// can be assigned
+				coverPerDriver[assignment.Worker.ID] = m.NewConstraint(mip.LessThanOrEqual, 1.0)
+				coverPerDriver[assignment.Worker.ID].NewTerm(1.0, x.Get(assignment))
+			} else {
+				constraint.NewTerm(1.0, x.Get(assignment))
```

Comment:
Does that actually work? I guess 🤔
Maybe better to do

```go
constraint, ok := coverPerDriver[assignment.Worker.ID]
if !ok {
    constraint = m.NewConstraint(mip.LessThanOrEqual, 1.0)
    coverPerDriver[assignment.Worker.ID] = constraint
}
constraint.NewTerm(1.0, x.Get(assignment))


```

Reply:
results in the same. fine with me